index.js

```diff
@@ -0,0 +1,241 @@
+      res.statusCode = 401;
+      res.setHeader('WWW-Authenticate', 'Basic realm="Incorrect Credentials"');
+      res.end('Access denied');
+      // this will prompt users one more time for the correct password,
+      // then display the login page WITHOUT saving the basic auth credentials.
+      // if they hit login, it'll show the auth form again, but won't show it a
+      // second time if they enter the wrong info.
+      // note: if the user enters the correct info on the second form,
+      // it'll show the login page but they'll have to click login again
+      // (and it'll automatically log them in without having to re-enter credentials)
+      // note: all of the above is the default behavior in amphora, but we're
+      // going to use varnish to automatically redirect them back to the ldap auth
+    } else {
+      res.send(tpl({
+        path: getPathOrBase(site),
+        flash: flash,
+        currentProviders: currentProviders,
+        user: req.user,
+        logoutLink: `${getAuthUrl(site)}/logout`
+      }));
+    }
+  };
+}
+
+/**
+ * middleware to log out. redirects to login page
+ * note: it goes to login page because usually users are navigating from edit mode,
+ * and they can't be redirected back into edit mode without logging in
+ * @param {object} site
+ * @returns {function}
+ */
+function onLogout(site) {
+  return function (req, res) {
+    req.logout();
+    res.redirect(`${getAuthUrl(site)}/login`);
+  };
+}
+
+/**
+ * There exists a case in which a user has an active session and
+ * is then removed as a user from a Clay instance. We must handle
+ * the error by re-directing the user to the login page and logging
+ * them out of their current session
+ *
+ * @param  {Object} site
+ * @returns {Function}
+ */
+function checkAuthentication(site) {
+  return function (err, req, res, next) {
+    if (err) {
+      onLogout(site)(req, res);
+    } else {
+      next();
+    }
+  };
+}
+
+/**
+ * initialize authentication
+ * @param {express.Router} router
+ * @param {array} providers (may be empty array)
+ * @param {object} site config for the site
+ * @param {object} storage
+ * @returns {array}
+ */
+function init(router, providers, site, storage) {
+  if (_isEmpty(providers)) {
+    return []; // exit early if no providers are passed in
+  }
+
+  setDb(storage);
+
+  const tpl = compileLoginPage('login.handlebars'),
+    currentProviders = getProviders(providers, site);
+
+  createStrategy(providers, site); // allow mocking this in tests
+
+  // init session authentication
+  router.use(sessionStore());
+  router.use(passport.initialize());
+  router.use(passport.session());
+  router.use(flash());
+
+  // protect routes
+  router.use(protectRoutes(site));
+
+  // add authorization routes
+  // note: these (and the provider routes) are added here,
+  // rather than as route controllers in lib/routes/
+  router.get('/_auth/login', onLogin(tpl, site, currentProviders));
```

Comment:
I have a question, is it possible to `const tpl = compileLoginPage('login.handlebars')` inside `onLogin` method? because I see it's only been used there.